skyverge/wc-plugin-framework
# Support cacheable requests in API base class

## Summary

This PR adds request/response caching support to the API base class in a transparent manner. This allows caching requests on a per-request basis. Meaning, if a request class declares support for caching, it's also possible to disable request caching for a concrete request by setting using the `$request->set_force_refresh( true )` setter, before performing the request.

### Story: [MWC-2478](https://jira.godaddy.com/browse/MWC-2478)

## Details

Originally, I had anticipated that the `CacheableRequestTrait` would be responsible for generating the transient (cache) key for the request. However, as the request instance does not have access to the plugin instance, I decided that the transient key should be generated in the `SV_WC_API_Base` class instead.

On a related note - I also did not add a way to filter or override the params which are used to generate the transient key at the moment. I could not think of an immediate use-case for this. Also, it allowed to keep the implementation much simpler for now, since we're just building the key base don the request URI and body. However, I'm open to suggestions.

The overall caching implementation is such that it should not affect any existing plugins in any way. Caching is completely opt-in. 

The raw response is stored in the cache, as this allows normal response handling and broadcasting. However, the raw response will only be stored after it's been successfully parsed. This will prevent storing the response if the request failed or if there was a parsing error. Consequently, it will also protect from corrupted transients, as the response handler will fail with an error, if the cached response data is broken.

Note that a `response_loaded_from_cache` flag is stored on the API instance, which should help debugging.

## Example Usage

```php

class ExchangeRatesRequest extends SV_WC_API_JSON_Request {

    use CacheableRequestTrait;

    public function __construct() {
        $this->cache_lifetime = 600; // override the deafult cache lifetime
    }

    // ...

}

class ExchangeRateAPI extends SV_WC_API_Base {

    public function get_currencies( $refresh = false )
    {
        $request = new ExchangeRatesRequest();

        if ( $refresh ) {
            $request->set_force_refresh( true );
        }

        $this->perform_request( $request );
    }

    // ..
}

$api = new ExchangeRateAPI();

$api->get_currencies(); // this should be a fresh request, the response will be cached
$api->get_currencies(); // this should use the cached response
$api->get_currencies( true ); // this should be a fresh request again, the response will be cached

// override cache timeout on runtime:
$request->set_cache_lifetime( 60 ); // note - this won't invalidate a previous cache with a longer lifetime - should it?

```

### Code style considerations

See my comments in #552

## Open questions

- [x] Should changing the cache lifetime invalidate any previously cached responses? Ie, should the lifetime be part of the cache key MD5 hash?

## QA

- [x] Review and merge #552 first
- [x] Change the base branch to `release/5.10.10`
- [x] Unit & integration tests pass `vendor/bin/codecept run unit && vendor/bin/codecept run integration`

Note: it probably makes sense to review this in tandem with https://github.com/gdcorp-partners/woocommerce-avatax/pull/74 as it helps to validate the functionality and ensure code is working as expected (we don't have tests for the API base class 😬 )
